lib/ddtrace/transport/io/traces.rb

```diff
@@ -0,0 +1,48 @@
+require 'ddtrace/transport/traces'
+
+require 'ddtrace/transport/io/response'
+require 'ddtrace/transport/io/client'
+
+module Datadog
+  module Transport
+    module IO
+      # IO transport behavior for traces
+      module Traces
+        # Response from HTTP transport for traces
+        class Response < IO::Response
+          include Transport::Traces::Response
+        end
+
+        # Extensions for HTTP client
+        module Client
+          def send_traces(traces)
+            # Build a request
+            req = Transport::Traces::Request.new(traces)
+
+            send_request(req) do |out, request|
+              # Encode trace data
+              encoded_data = request.parcel.encode_with(encoder)
```

Comment:
Looking at our current `Transport::HTTP::Traces` module, I see that its only responsibility is to convert the `tracers` argument into a `Request` object, and forward it down the transport chain.

Here in `Transport::IO::Traces` we have encoding logic, as well as "adapter" logic (`out.puts(data)`).

I understand that this transport is much simpler than our current HTTP transport, meaning we'd have fewer modules doing more work, but it seems like, similar to what @brettlangdon mentioned above, that there's duplicate logic around encoding and adapter logic among this `Traces::Client` module and `IO::Client`.

I wasn't able to find similar "duplication" in the HTTP implementation, but again, the HTTP implementation has Spec and API abstractions that we don't really need here, so it's not a necessarily a strict comparison.

Reply:
They aren't duplicates, they are separate functions with different use cases; the encoding behavior looks similar, but they are not the same. One calls `#encode`, the other calls `#encode_traces`, one builds an `IO::Response`, the other builds a `IO::Traces::Response`.

I deliberately removed adapters and a lot of the other bells & whistles that came with `HTTP`; the idea is that `Traces` is not an adapter, just a "helper". It's the same for `HTTP` as well.

I'm taking another look to see if I can consolidate the encoding behavior a little more, but otherwise I don't see the duplication.

Reply:
See https://github.com/DataDog/dd-trace-rb/pull/910#discussion_r368113634